settings: Add setting to control how animated images are played.

roanster007 opened the pull request:
<!-- Describe your pull request here.-->

Previously animated images are automatically played in the message feed of the web app.

This commit adds a new personal setting -- "web_animate_image_previews" which controls how the animated images would be played in the web app message feed -- always played, on hover, or only in the image viewer.

Fixes #31016

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

<details><summary>Screenshots and screen captures:</summary>
<p>

![image](https://github.com/user-attachments/assets/f0e4c9c2-d355-446a-b3a8-e48c761ce854)

</p>
</details> 


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

roanster007 commented:
@zulipbot add "buddy review"

roanster007 commented:
@Vector73 could you review this (only the last commit).

timabbott commented on web/src/lightbox.ts:
```diff
@@ -580,7 +580,7 @@ export function initialize(): void {
     $("#main_div, #compose .preview_content").on(
         "click",
-        ".message_inline_image:not(.message_inline_video) a",
+        ".message_inline_image:not(.message_inline_video) a, .message_inline_animated_image",
```
Should this have a ` a` in the selector?

roanster007 replied:
No, then you won't be able to click on the play icon so as to open it in lightbox (other cases are anyways handled by `.message_inline_image:not(.message_inline_video) a`).

timabbott replied:
OK

timabbott commented on web/src/rendered_markdown.ts:
```diff
@@ -346,7 +345,13 @@ export const update_elements = ($content: JQuery): void => {
+                    $inline_img_thumbnail
+                        .closest(".message_inline_image")
+                        .addClass("message_inline_animated_image");
+                }
```
Why the name `message_inline_animated_image`? It seems like this is used for the case where the thing is not currently animated, but can be animated.

roanster007 replied:
basically it would be added when the animated images are is not currently playing i.e., to get the play icon on the image. So, yeah it is definitely not a good name. So I guess I'll update it to `message_inline_animated_image_paused`.

timabbott replied:
Yeah, or `message_inline_animated_image_still` might be even better, since `still` is the term we use for the non-animated preview files for animated images.

timabbott commented on web/src/server_events_dispatch.js:
```diff
@@ -779,6 +780,9 @@ export function dispatch_normal_event(event) {
             }
+            if (event.property === "web_animate_image_previews") {
+                message_lists.current?.rerender();
+            }
```
I suspect you need to do this using `all_rendered_message_lists`, to make sure the "combined feed" gets rerendered if it is cached. .

roanster007 replied:
updated!

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -11334,6 +11334,23 @@ paths:
                   example: 1
+                web_animate_image_previews:
+                  description: |
+                    Controls how animated images should be played in the message feed.
```
"in the web/desktop application" should be in these descriptions.

roanster007 replied:
updated!

timabbott commented:
Code looks basically fine to me; I've not tested yet. Posted a handful of comments above.

alya commented:
Why does the "play" button look different between an animated image and a video file?

alya commented:
Works for me in manual testing! (I haven't tested on a mobile device.)

alya commented:
I think it would be better if spectator got the default user settings experience; currently, it looks like it's always the on-hover behavior. But this is not critical if it's not an easy thing to do.

timabbott commented:
Rebased this and I updated the test so it passed; haven't tried to resolve the above feedback yet.

roanster007 commented:
Addressed all reviews except:
> I think it would be better if spectators got the default user settings experience; currently, it looks like it's always the on-hover behavior. But this can be a follow-up if it's not an easy thing to do.

Since the default option for the user setting is on_hover, I guess for spectators we want the option to be always?

timabbott commented on web/src/rendered_markdown.ts:
```diff
@@ -345,7 +345,13 @@ export const update_elements = ($content: JQuery): void => {
+                } else {
+                    $inline_img_thumbnail
+                        .closest(".message_inline_image")
+                        .addClass("message_inline_animated_image_still");
```
Added the requested behavior of treating "hover" as "always" on mobile devices.

``` diff
diff --git a/web/src/rendered_markdown.ts b/web/src/rendered_markdown.ts
index aa19e7cf69..9081b7e7c9 100644
--- a/web/src/rendered_markdown.ts
+++ b/web/src/rendered_markdown.ts
@@ -345,9 +345,18 @@ export const update_elements = ($content: JQuery): void => {
             const $inline_img_thumbnail = $(this);
             let thumbnail_name = thumbnail.preferred_format.name;
             if ($inline_img_thumbnail.attr("data-animated") === "true") {
-                if (user_settings.web_animate_image_previews === "always") {
+                if (user_settings.web_animate_image_previews === "always" ||
+                    (user_settings.web_animate_image_previews === "on_hover" &&
+                        util.is_mobile())) {
+                    // Use a constantly animating thumbnail in the
+                    // "always" configuration, or the
+                    // "on_hover" configuration on mobile devices,
+                    // where hovering is impossible and there's much
+                    // less on the screen.
                     thumbnail_name = thumbnail.animated_format.name;
                 } else {
+                    // If we're showing a still thumbnail, show a play
+                    // button so that users that it can be played.
                     $inline_img_thumbnail
                         .closest(".message_inline_image")
                         .addClass("message_inline_animated_image_still");
```

timabbott commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,14 @@ format used by the Zulip server that they are interacting with.
+  /settings`](/api/update-settings), [`PATCH
+  /realm/user_settings_defaults`](/api/update-realm-user-settings-defaults):
+  Added new `web_animate_image_previews` setting, which controls how
+  animated images should be played in the message feed.
```
I make this mention "web/desktop" as well.

timabbott commented:
Pushed some small changes noted above and marked this to merge once CI passes, thanks for making this happen in time for 9.0 @roanster007!

timabbott commented:
Apparently I didn't actually mark it to merge :)
